src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Match.cs

```diff
@@ -81,9 +81,10 @@ internal void Reset(Regex regex, string text, int textbeg, int textend, int text
             _textend = textend;
             _textstart = textstart;
 
-            for (int i = 0; i < _matchcount.Length; i++)
+            int[] matchcount = _matchcount;
```

Comment:
This kind of optimization surprises me. I would have thought there to be no perf difference between these two patterns. Is it because the field isn't `readonly`, and therefore needs to be looked up each time through the loop?

Reply:
When it comes to optimizations, the JIT has to assume that the field changes between accesses (even if it's readonly, since the instance field can still be set via reflection) and thus can't do as well with things like bounds check elimination, e.g.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAMyUATOS7kA3svJHyfKKwnYMMcqxYBtALrkA+tihRFS4+UPG7AWRgMAAsIABMASX5BAApAkPCovkEAeT42CCZcGgBBAHM82Fxccxg0tnxWAC9LVkyASgdfI3VKFHJ/Khj6/WbvcgAzaHIY2wwbcgBeclI5CYAeFzcoGgAZGCY8kLnWAGpd+r7+41d3O1Ynadm+gF8+voCg0MjouKfE6PK6rNyCopKJGV0uxqrUGk0vMZWsR2v5NN1epDjmNHORllMlu5PMcjEMoCMxhMrjtyItlmsNltgjt9ockTijMtzpcZtjjHclDcgA=